File: src/novel_manga/application/dashboard/workbench.py
"""Book discovery and content reads for the workbench pages.

The board pages track the configured production novels.  The workbench discovers every book on
disk - an outputs/<dir> holding a profile.json is a book, whether production manages it or not -
and reads the artifacts an episode directory already holds: the script, the prompts, the plans,
the reviews and the rendered film itself.  Everything here is read-only.
"""
from __future__ import annotations
import json
import os
import re
import threading
import time
from pathlib import Path

from novel_manga.application.configuration import pipeline_config

MEDIA_TYPES = {".mp4": "video/mp4", ".jpeg": "image/jpeg", ".jpg": "image/jpeg",
               ".png": "image/png", ".webp": "image/webp",
               ".wav": "audio/wav", ".m4a": "audio/mp4", ".mp3": "audio/mpeg"}

TEXT_CAP = 400_000      # a text artifact larger than this is cut, with its full size noted
LOG_TAIL = 200          # log files ship their last lines, not the whole run
BOOK_ID = re.compile(r"[A-Za-z0-9._-]+")

_cache: dict = {}
_cache_lock = threading.Lock()
CACHE_SECONDS = 30


def _cached(key, build):
    with _cache_lock:
        hit = _cache.get(key)
        if hit and time.monotonic() - hit[0] < CACHE_SECONDS:
            return hit[1]
    data = build()
    with _cache_lock:
        _cache[key] = (time.monotonic(), data)
    return data


def _read_json(path: Path):
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None


def _read_text(path: Path, cap: int = TEXT_CAP):
    try:
        text = path.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return None
    size = path.stat().st_size
    if len(text) > cap:
        return {"text": text[:cap], "truncated": True, "size": size}
    return {"text": text, "truncated": False, "size": size}


def _read_tail(path: Path, lines: int = LOG_TAIL):
    read = _read_text(path, cap=600_000)
    if read is None:
        return None
    tail = read["text"].splitlines()[-lines:]
    return {"text": "\n".join(tail), "tail_of": read["size"]}


def book_dir(root, book_id: str) -> Path:
    """The outputs directory of a discovered book; anything else is not addressed by these pages."""
    if not BOOK_ID.fullmatch(book_id):
        raise KeyError(book_id)
    directory = Path(root) / "outputs" / book_id
    if not (directory / "profile.json").is_file():
        raise KeyError(book_id)
    return directory


def _episode_dirs(directory: Path) -> list[Path]:
    try:
        with os.scandir(directory) as entries:
            return sorted((Path(e.path) for e in entries if e.is_dir()
                           and e.name.startswith(directory.name + "_")
                           and e.name.rsplit("_", 1)[-1].isdigit()),
                          key=lambda p: int(p.name.rsplit("_", 1)[-1]))
    except OSError:
        return []


def books(root) -> dict:
    """Every book on disk: the pipeline's managed ones first, then anything with a profile.json."""
    root = Path(root)
    configured = pipeline_config(root).get("novels", [])
    indexed = sorted(enumerate(configured), key=lambda item: item[1].get("display_order", item[0]))
    order = {n["id"]: i for i, (_, n) in enumerate(indexed)}
    titles = {n["id"]: n.get("title", n["id"]) for n in configured}
    outputs = root / "outputs"
    rows = []
    try:
        with os.scandir(outputs) as entries:
            dirs = sorted(Path(e.path) for e in entries if e.is_dir() and not e.name.startswith("_"))
    except OSError:
        dirs = []
    for directory in dirs:
        if not (directory / "profile.json").is_file():
            continue
        book_id = directory.name
        profile = _read_json(directory / "profile.json") or {}
        novel = _read_json(directory / "novel.json") or {}
        delivery = _read_json(directory / "delivery.json") or {}
        assets_dir = directory / "series_assets"
        rows.append({
            "id": book_id,
            "title": novel.get("title") or titles.get(book_id) or book_id,
            "managed": book_id in titles,
            "backend": profile.get("planning_backend", "local"),
            "style": profile.get("style"), "frame": profile.get("frame"), "genre": profile.get("genre"),
            "episodes": len(_episode_dirs(directory)),
            "deliverable": delivery.get("deliverable"), "total": delivery.get("total"),
            "has_assets": assets_dir.is_dir(),
        })
    rows.sort(key=lambda r: (r["id"] not in titles, order.get(r["id"], 999), r["id"]))
    return {"now": time.strftime("%F %T"), "books": rows}


def episodes(root, book_id: str) -> dict:
    directory = book_dir(root, book_id)

    def build():
        rows = []
        for episode_dir in _episode_dirs(directory):
            try:
                with os.scandir(episode_dir) as entries:
                    names = {e.name for e in entries}
            except OSError:
                continue
            number = int(episode_dir.name.rsplit("_", 1)[-1])
            rows.append({
                "episode": number,
                "video": f"{episode_dir.name}.mp4" in names,
                "script": "chapter_script.md" in names,
                "plan": "clip_plan.md" in names,
                "review": "episode_review.json" in names,
                "qc": "media_qc_report.json" in names,
                "repair_candidate": "repair_candidate" in names,
                "mtime": episode_dir.stat().st_mtime,
            })
        return {"book": book_id, "episodes": rows}

    return _cached(("episodes", str(directory)), build)


def _clip_rows(plan: dict | None) -> list[dict]:
    """One row per clip for the jump list: index, kind, seconds and a short label."""
    if not isinstance(plan, dict):
        return []
    rows = []
    offset = 0.0
    for index, clip in enumerate(plan.get("clips") or []):
        if not isinstance(clip, dict):
            continue
        seconds = clip.get("request_seconds") or clip.get("estimated_seconds") or 0
        label = ""
        for field in ("text", "prompt", "event", "narration", "description"):
            if clip.get(field):
                label = str(clip[field])[:80]
                break
        rows.append({"n": index + 1, "kind": clip.get("kind"), "seconds": seconds,
                     "offset": round(offset, 2), "label": label})
        offset += float(seconds or 0)
    return rows


def _agent_storyboard(directory: Path) -> dict | None:
    """The sandbox agent's own storyboard for this episode, when one was imported: the import
    record, and the sheet itself read by the same workbook reader the import uses."""
    record_path = directory / "agent_storyboard.json"
    if not record_path.is_file():
        return None
    from novel_manga.planning.storyboard import read_workbook
    result = {"record": _read_json(record_path) or {}, "sheets": []}
    sheet_dir = directory / "agent_storyboard"
    if sheet_dir.is_dir():
        for xlsx in sorted(sheet_dir.glob("*.xlsx")):
            try:
                for sheet in read_workbook(xlsx):
                    result["sheets"].append({"name": sheet.name, "file": xlsx.name,
                                             "notes": list(sheet.notes),
                                             "rows": [dict(row) for row in sheet.rows]})
            except Exception as error:  # noqa: BLE001 - a broken sheet must not break the page
                result["sheets"].append({"name": xlsx.stem, "file": xlsx.name, "error": str(error)})
    return result


def episode(root, book_id: str, number: int) -> dict:
    directory = book_dir(root, book_id) / f"{book_id}_{number}"
    if not directory.is_dir():
        raise KeyError(number)
    name = directory.name
    profile = _read_json(directory.parent / "profile.json") or {}
    plan = _read_json(directory / "clip_plan.json")
    prompts = []
    for path in sorted(directory.glob("request_attempt_*.json")) + sorted(directory.glob("analysis_attempt_*.txt")):
        body = _read_json(path) if path.suffix == ".json" else None
        prompts.append({"name": path.name,
                        "json": body if body is not None else None,
                        "text": None if body is not None else _read_text(path)})
    logs = {}
    for log in ("plan.log", "render.log", "h3_prompts.log"):
        tail = _read_tail(directory / log)
        if tail is not None:
            logs[log] = tail
    media = lambda fn: f"{name}/{fn}" if (directory / fn).is_file() else None
    previous = directory / "repair_history" / "previous_final.mp4"
    return {
        "book": book_id, "episode": number,
        "backend": profile.get("planning_backend", "local"),
        "agent_storyboard": _agent_storyboard(directory),
        "video": media(f"{name}.mp4"),
        "cover": media(f"{name}_cover.jpeg"),
        "ending": media(f"{name}_ending.jpeg"),
        "previous_video": f"{name}/repair_history/previous_final.mp4" if previous.is_file() else None,
        "script_md": _read_text(directory / "chapter_script.md"),
        "plan_md": _read_text(directory / "clip_plan.md"),
        "clips": _clip_rows(plan),
        "review": _read_json(directory / "episode_review.json"),
        "qc": _read_json(directory / "media_qc_report.json"),
        "report": _read_json(directory / "chapter_script_report.json"),
        "prompts": prompts,
        "logs": logs,
    }


def _asset_dirs(directory: Path) -> list[dict]:
    rows = []
    try:
        with os.scandir(directory) as entries:
            dirs = sorted(Path(e.path) for e in entries if e.is_dir())
    except OSError:
        return rows
    for asset in dirs:
        spec = _read_json(asset / "spec.json") or {}
        try:
            with os.scandir(asset) as entries:
                images = sorted(e.name for e in entries
                                if e.is_file() and Path(e.name).suffix.lower() in MEDIA_TYPES
                                and MEDIA_TYPES[Path(e.name).suffix.lower()].startswith("image/"))
        except OSError:
            images = []
        rows.append({"id": asset.name, "spec": spec, "images": images})
    return rows


def assets(root, book_id: str) -> dict:
    directory = book_dir(root, book_id) / "series_assets"
    if not directory.is_dir():
        return {"book": book_id, "characters": [], "locations": [], "voices": [], "avatars": []}

    def flat(kind: str) -> list[dict]:
        """Media files that sit directly under the kind directory (voices, avatars)."""
        target = directory / kind
        rows = []
        try:
            with os.scandir(target) as entries:
                for e in sorted(entries, key=lambda x: x.name):
                    suffix = Path(e.name).suffix.lower()
                    if e.is_file() and suffix in MEDIA_TYPES:
                        rows.append({"name": e.name, "type": MEDIA_TYPES[suffix]})
        except OSError:
            pass
        return rows

    return _cached(("assets", str(directory)), lambda: {
        "book": book_id,
        "characters": _asset_dirs(directory / "characters"),
        "locations": _asset_dirs(directory / "locations"),
        "voices": flat("voices"),
        "avatars": flat("avatars"),
    })


def experiments(root) -> dict:
    """The research shelf: every outputs/experiments/<dir>, plus which books an agent plans.

    An experiment registers itself with a manifest.json (the benchmark harness already writes
    one); a directory without one is still listed, marked 未登记, so work done in another session
    shows up either way.  Status is inferred from what landed: a final summary means finished,
    recent file activity means running.
    """
    root = Path(root)
    directory = root / "outputs" / "experiments"
    rows = []
    try:
        with os.scandir(directory) as entries:
            dirs = sorted(Path(e.path) for e in entries if e.is_dir())
    except OSError:
        dirs = []
    for experiment in dirs:
        manifest = _read_json(experiment / "manifest.json") or {}
        results, latest = [], 0.0
        try:
            with os.scandir(experiment) as entries:
                for e in entries:
                    latest = max(latest, e.stat().st_mtime)
                    if e.name == "manifest.json":
                        continue
                    if (e.is_file() and (e.name.endswith(".json") or e.name == "report.md")) or e.name == "evaluation":
                        results.append(e.name)
        except OSError:
            pass
        results = sorted(results)[:8]
        if (experiment / "final-summary.json").is_file():
            status = "完成"
        elif latest and time.time() - latest < 2 * 3600:
            status = "活跃"
        else:
            status = "存档"
        rows.append({
            "name": experiment.name,
            "registered": bool(manifest),
            "created_at": manifest.get("created_at"),
            "comparison": manifest.get("comparison"),
            "model": manifest.get("model"), "head": manifest.get("head"),
            "endpoints": manifest.get("endpoints"),
            "status": status, "results": results,
            "latest": latest,
        })
    rows.sort(key=lambda r: r["latest"], reverse=True)

    shelf = root / "outputs" / "agent-test"
    agent_test = []
    try:
        with os.scandir(shelf) as entries:
            for e in sorted(entries, key=lambda x: x.name):
                if e.is_dir():
                    with os.scandir(e.path) as inner:
                        count = sum(1 for _ in inner)
                    agent_test.append({"name": e.name, "files": count})
                elif e.is_file():
                    agent_test.append({"name": e.name, "files": None})
    except OSError:
        pass
    agent_books = [{"id": b["id"], "title": b["title"], "backend": b["backend"]}
                   for b in books(root)["books"] if b["backend"] != "local"]
    return {"now": time.strftime("%F %T"), "experiments": rows,
            "agents": {"books": agent_books, "agent_test": agent_test}}


def media_file(root, book_id: str, relative: str) -> Path:
    """A whitelisted media file inside the book's own directory; nothing else is served."""
    directory = book_dir(root, book_id).resolve()
    target = (directory / relative).resolve()
    if not target.is_relative_to(directory) or target.suffix.lower() not in MEDIA_TYPES or not target.is_file():
        raise KeyError(relative)
    return target


def thumbnail(root, book_id: str, relative: str, width: int = 520) -> Path:
    """A downscaled JPEG of an image asset, cached under outputs/.dashboard/thumbs.

    Cards are 3-5 MB at 1152x2048 and a gallery page shows hundreds; the browser only needs a
    few hundred pixels.  The cache key carries the source mtime, so a redrawn card earns a new
    thumbnail, and the file lands by rename so a half-written one is never served.
    """
    source = media_file(root, book_id, relative)
    if not MEDIA_TYPES[source.suffix.lower()].startswith("image/"):
        raise KeyError(relative)
    width = max(64, min(int(width), 1600))
    stat = source.stat()
    import hashlib
    key = hashlib.sha1(f"{book_id}/{relative}:{stat.st_mtime_ns}:{width}".encode()).hexdigest()[:20]
    target = Path(root) / "outputs" / ".dashboard" / "thumbs" / book_id / f"{key}.jpg"
    if target.is_file():
        return target
    from PIL import Image
    with Image.open(source) as image:
        image = image.convert("RGB")
        if image.width > width:
            image = image.resize((width, round(image.height * width / image.width)), Image.LANCZOS)
        target.parent.mkdir(parents=True, exist_ok=True)
        staging = target.with_suffix(".tmp")
        image.save(staging, "JPEG", quality=82)
        os.replace(staging, target)
    return target

File: src/novel_manga/dashboard/static/workbench.js
/* Workbench pages: /workbench (books), /novel/<id> (episodes), /episode/<id>/<n> (detail),
   /assets/<id> (gallery).  All reads come from /api/book/... and /media/<book>/... . */

const esc = s => String(s ?? "").replace(/[&<>"']/g, c => ({"&":"&amp;","<":"&lt;",">":"&gt;",'"':"&quot;","'":"&#39;"}[c]));
const parts = location.pathname.split("/").filter(Boolean);
const fmtBytes = n => n > 1048576 ? (n/1048576).toFixed(1)+" MB" : n > 1024 ? (n/1024).toFixed(0)+" KB" : n+" B";
const fmtTime = t => { const d = new Date(t*1000); return `${d.getFullYear()}-${String(d.getMonth()+1).padStart(2,"0")}-${String(d.getDate()).padStart(2,"0")} ${String(d.getHours()).padStart(2,"0")}:${String(d.getMinutes()).padStart(2,"0")}`; };

async function getJSON(url){
  const r = await fetch(url);
  if (!r.ok) throw new Error(url + " → " + r.status);
  return r.json();
}

function badge(ok, label){ return ok ? `<span class="pill ok-p">${label}</span>` : `<span class="pill off">${label}</span>`; }

/* ---------- /workbench ---------- */
async function renderBooks(){
  const data = await getJSON("/api/workbench");
  const books = data.books;
  document.getElementById("stats").innerHTML =
    stat("磁盘上的书", books.length, "本") +
    stat("生产纳管", books.filter(b=>b.managed).length, "本") +
    stat("未纳管", books.filter(b=>!b.managed).length, "本", "没在 configs/pipeline.json 里，只在磁盘上") +
    stat("剧集合计", books.reduce((s,b)=>s+b.episodes,0), "集");
  document.getElementById("books").innerHTML =
    `<tr><th>书</th><th>状态</th><th>画风/画幅</th><th>剧集</th><th>可交付</th><th></th></tr>` +
    books.map(b=>`<tr>
      <td><b>${esc(b.title)}</b> <span class="dim">${esc(b.id)}</span></td>
      <td>${b.managed ? '<span class="pill ok-p">纳管</span>' : '<span class="pill warn-p">未纳管</span>'}
        ${b.backend && b.backend!=="local" ? `<span class="pill warn-p" title="planning_backend">${esc(b.backend)}</span>` : ""}</td>
      <td class="dim">${esc(b.style||"—")} · ${esc(b.frame||"—")}</td>
      <td class="num">${b.episodes}</td>
      <td>${b.deliverable==null ? '<span class="dim">未统计</span>' : `<b class="num">${b.deliverable}</b> / ${b.total}`}</td>
      <td class="links"><a href="/novel/${encodeURIComponent(b.id)}">剧集</a>
        ${b.has_assets ? `<a href="/assets/${encodeURIComponent(b.id)}">资产</a>` : ""}</td>
    </tr>`).join("");
}
function stat(k, v, u, sub){
  return `<div class="stat"><div class="k">${k}</div><b>${v}</b><span class="u">${u}</span>${sub?`<div class="sub">${sub}</div>`:""}</div>`;
}

/* ---------- /novel/<id> ---------- */
async function renderNovel(book){
  const data = await getJSON(`/api/book/${encodeURIComponent(book)}/episodes`);
  const rows = data.episodes;
  const videos = rows.filter(r=>r.video).length;
  document.getElementById("book-head").innerHTML = `<div class="stats">
    ${stat("剧集", rows.length, "集")}${stat("有成片", videos, "集")}${stat("有剧本", rows.filter(r=>r.script).length, "集")}${stat("修复候选", rows.filter(r=>r.repair_candidate).length, "集")}
    </div>
    <div class="dim" style="margin:4px 2px 0">书 <b>${esc(book)}</b> · <a href="/assets/${encodeURIComponent(book)}">资产库 →</a></div>`;
  document.getElementById("episodes").innerHTML =
    `<tr><th>集</th><th>成片</th><th>剧本</th><th>分镜</th><th>审片</th><th>质检</th><th>修复</th><th>更新时间</th></tr>` +
    rows.map(r=>`<tr>
      <td><a href="/episode/${encodeURIComponent(book)}/${r.episode}"><b class="num">${r.episode}</b></a></td>
      <td>${badge(r.video,"成片")}</td><td>${badge(r.script,"剧本")}</td><td>${badge(r.plan,"分镜")}</td>
      <td>${badge(r.review,"审片")}</td><td>${badge(r.qc,"质检")}</td><td>${r.repair_candidate?'<span class="pill warn-p">候选</span>':""}</td>
      <td class="dim num">${r.mtime?fmtTime(r.mtime):""}</td>
    </tr>`).join("");
}

/* ---------- /episode/<id>/<n> ---------- */
function textBlock(title, body){
  if (!body) return "";
  const note = body.truncated ? `<span class="dim">（截断展示，完整 ${fmtBytes(body.size)}）</span>` : "";
  return `<div class="card"><div class="label">${title} ${note}</div><pre class="md">${esc(body.text)}</pre></div>`;
}
function jsonBlock(title, obj){
  if (obj == null) return "";
  return `<details class="card"><summary class="label">${title}</summary><pre class="md">${esc(JSON.stringify(obj,null,2))}</pre></details>`;
}
async function renderEpisode(book, n){
  const d = await getJSON(`/api/book/${encodeURIComponent(book)}/episode/${n}`);
  const media = rel => `/media/${encodeURIComponent(book)}/${rel.split("/").map(encodeURIComponent).join("/")}`;
  const thumb = rel => `/thumb/${encodeURIComponent(book)}/${rel.split("/").map(encodeURIComponent).join("/")}`;
  let html = `<div class="dim" style="margin:0 2px"><a href="/novel/${encodeURIComponent(book)}">← ${esc(book)}</a> · 第 <b>${n}</b> 集
    ${d.report && d.report.model ? `· 规划模型 <span class="pill">${esc(d.report.model)}</span>` : ""}</div>`;
  if (d.video){
    const clipRows = (d.clips||[]).map(c=>`<tr data-t="${c.offset}" class="cliprow"><td class="num">${c.n}</td><td class="dim">${esc(c.kind||"")}</td><td class="num">${c.seconds}s</td><td>${esc(c.label)}</td></tr>`).join("");
    html += `<div class="card"><div class="label">成片</div>
      <video id="film" controls preload="metadata" poster="${d.cover?thumb(d.cover)+"?w=960":""}" src="${media(d.video)}"></video>
      ${d.previous_video ? `<div class="dim" style="margin-top:6px">修复前版本：<a href="${media(d.previous_video)}" target="_blank">previous_final.mp4</a></div>` : ""}
      ${clipRows ? `<div class="twrap" style="margin-top:10px"><table class="resp"><tr><th>#</th><th>类型</th><th>时长</th><th>内容</th></tr>${clipRows}</table><div class="dim" style="margin-top:4px">点任意行跳到该片段起点</div></div>` : ""}
    </div>`;
  } else {
    html += `<div class="card"><div class="label">成片</div><div class="dim">还没有成片文件</div></div>`;
  }
  if (d.prompts && d.prompts.length){
    html += `<div class="card"><div class="label">提示词请求 · ${d.prompts.length} 次</div>` + d.prompts.map((p,i)=>{
      const body = p.json != null ? esc(JSON.stringify(p.json,null,2))
                 : p.text ? esc(p.text.text) + (p.text.truncated ? `\n…（截断，完整 ${fmtBytes(p.text.size)}）` : "") : "";
      return `<details><summary class="num">${esc(p.name)}</summary><pre class="md">${body}</pre></details>`;
    }).join("") + `</div>`;
  }
  html += textBlock("剧本 chapter_script.md", d.script_md);
  html += textBlock("分镜表 clip_plan.md", d.plan_md);
  html += jsonBlock("审片意见 episode_review.json", d.review);
  html += jsonBlock("媒体质检 media_qc_report.json", d.qc);
  html += jsonBlock("规划报告 chapter_script_report.json", d.report);
  const logKeys = Object.keys(d.logs||{});
  if (logKeys.length){
    html += `<div class="card"><div class="label">日志（各取末尾 200 行）</div>` +
      logKeys.map(k=>`<details><summary class="num">${esc(k)}</summary><pre class="md">${esc(d.logs[k].text)}</pre></details>`).join("") + `</div>`;
  }
  document.getElementById("episode").innerHTML = html;
  const film = document.getElementById("film");
  if (film) document.querySelectorAll(".cliprow").forEach(row =>
    row.addEventListener("click", ()=>{ film.currentTime = parseFloat(row.dataset.t)||0; film.play(); }));
}

/* ---------- /assets/<id> ---------- */
function specRows(spec){
  const skip = new Set(["asset_id"]);
  return Object.entries(spec||{}).filter(([k])=>!skip.has(k))
    .map(([k,v])=>`<tr><td class="dim" style="white-space:nowrap">${esc(k)}</td><td>${esc(typeof v==="object"?JSON.stringify(v):v)}</td></tr>`).join("");
}
function assetCard(book, kind, a){
  const spec = a.spec||{};
  const name = spec.name || spec.location || spec.title || a.id;
  const base = `/media/${encodeURIComponent(book)}/series_assets/${kind}/${a.id}`;
  const thumb = img => `/thumb/${encodeURIComponent(book)}/series_assets/${kind}/${a.id}/${encodeURIComponent(img)}`;
  const imgs = a.images.map(img=>({full:`${base}/${encodeURIComponent(img)}`, small:thumb(img)}));
  return `<div class="asset">
    ${imgs.length?`<a href="${imgs[0].full}" target="_blank"><img loading="lazy" src="${imgs[0].small}?w=520" alt="${esc(name)}"></a>`:""}
    <div class="asset-body">
      <div><b>${esc(name)}</b> ${spec.role?`<span class="pill">${esc(spec.role)}</span>`:""} <span class="dim num">${esc(a.id)}</span></div>
      ${spec.appearance?`<div class="dim small">${esc(spec.appearance)}</div>`:""}
      ${spec.wardrobe?`<div class="dim small">服装：${esc(spec.wardrobe)}</div>`:""}
      ${imgs.length>1?`<div class="thumbs">${imgs.slice(1).map(u=>`<a href="${u.full}" target="_blank"><img loading="lazy" src="${u.small}?w=240"></a>`).join("")}</div>`:""}
      <details><summary class="dim small">spec.json</summary><table class="kv">${specRows(spec)}</table></details>
    </div></div>`;
}
async function renderAssets(book){
  const d = await getJSON(`/api/book/${encodeURIComponent(book)}/assets`);
  let html = `<div class="dim" style="margin:0 2px"><a href="/novel/${encodeURIComponent(book)}">← ${esc(book)}</a> · 资产库</div>`;
  html += `<div class="stats">${stat("角色", d.characters.length, "张卡")}${stat("地点", d.locations.length, "张卡")}${stat("音色", d.voices.length, "条")}${stat("头像", d.avatars.length, "个")}</div>`;
  if (d.characters.length)
    html += `<div class="card"><div class="label">角色卡</div><div class="gallery">${d.characters.map(a=>assetCard(book,"characters",a)).join("")}</div></div>`;
  if (d.locations.length)
    html += `<div class="card"><div class="label">地点卡</div><div class="gallery">${d.locations.map(a=>assetCard(book,"locations",a)).join("")}</div></div>`;
  if (d.voices.length)
    html += `<div class="card"><div class="label">音色</div>` + d.voices.map(v=>
      `<div class="voice"><span class="num">${esc(v.name)}</span><audio controls preload="none" src="/media/${encodeURIComponent(book)}/series_assets/voices/${encodeURIComponent(v.name)}"></audio></div>`).join("") + `</div>`;
  if (d.avatars.length)
    html += `<div class="card"><div class="label">头像</div><div class="avatars">` + d.avatars.map(v=>
      `<a href="/media/${encodeURIComponent(book)}/series_assets/avatars/${encodeURIComponent(v.name)}" target="_blank"><img loading="lazy" src="/thumb/${encodeURIComponent(book)}/series_assets/avatars/${encodeURIComponent(v.name)}?w=160" title="${esc(v.name)}"></a>`).join("") + `</div></div>`;
  if (!d.characters.length && !d.locations.length && !d.voices.length && !d.avatars.length)
    html += `<div class="card"><div class="dim">series_assets 目录还没有资产</div></div>`;
  document.getElementById("assets").innerHTML = html;
}

/* ---------- /experiments ---------- */
const STATUS_PILL = {"完成": "ok-p", "活跃": "warn-p", "存档": "off"};
async function renderExperiments(){
  const d = await getJSON("/api/experiments");
  const agents = d.agents || {books: [], agent_test: []};
  let agentHtml = `<div class="card"><div class="label">Agent 运行情况</div>`;
  if (agents.books.length)
    agentHtml += `<div style="margin-bottom:10px">` + agents.books.map(b=>
      `<span class="pill warn-p" title="planning_backend=${esc(b.backend)}">${esc(b.title)} · ${esc(b.backend)}</span>`).join(" ") + `</div>`;
  else
    agentHtml += `<div class="dim" style="margin-bottom:10px">当前没有书使用非本地规划后端（profile.json 的 planning_backend）。</div>`;
  if (agents.agent_test.length)
    agentHtml += `<div class="label">沙箱 agent 产物 · outputs/agent-test</div><div class="twrap"><table class="resp">
      <tr><th>目录/文件</th><th>内容数</th></tr>` +
      agents.agent_test.map(t=>`<tr><td class="num">${esc(t.name)}</td><td class="num">${t.files==null?"文件":t.files}</td></tr>`).join("") +
      `</table></div>`;
  agentHtml += `</div>`;
  document.getElementById("agents").innerHTML = agentHtml;

  const rows = d.experiments;
  document.getElementById("experiments").innerHTML = rows.length ? rows.map(e=>`
    <details class="exp">
      <summary>
        <span class="pill ${STATUS_PILL[e.status]||"off"}">${e.status}</span>
        ${e.registered ? "" : '<span class="pill off">未登记</span>'}
        <b>${esc(e.name)}</b>
        <span class="dim">${e.latest?fmtTime(e.latest):""}</span>
      </summary>
      <div class="exp-body">
        ${e.comparison ? `<div>${esc(e.comparison)}</div>` : ""}
        <div class="dim small">
          ${e.model ? `模型 ${esc(e.model)} · ` : ""}${e.head ? `冻结于 ${esc(String(e.head).slice(0,10))} · ` : ""}${e.created_at ? `建于 ${esc(e.created_at)}` : ""}
        </div>
        ${e.endpoints ? `<div class="dim small">端点：${esc(e.endpoints.join("、"))}</div>` : ""}
        ${e.results.length ? `<div class="small">产物：${e.results.map(r=>`<span class="pill">${esc(r)}</span>`).join(" ")}</div>` : '<div class="dim small">还没有结果文件</div>'}
      </div>
    </details>`).join("") : `<div class="dim">outputs/experiments 还没有实验目录</div>`;
}

/* ---------- dispatch ---------- */
(async ()=>{
  try{
    if (parts[0]==="workbench") await renderBooks();
    else if (parts[0]==="novel" && parts[1]) await renderNovel(decodeURIComponent(parts[1]));
    else if (parts[0]==="episode" && parts[1] && parts[2]) await renderEpisode(decodeURIComponent(parts[1]), parts[2]);
    else if (parts[0]==="assets" && parts[1]) await renderAssets(decodeURIComponent(parts[1]));
    else if (parts[0]==="experiments") await renderExperiments();
  }catch(e){
    document.querySelector("main").insertAdjacentHTML("afterbegin",
      `<div class="card"><div class="label">加载失败</div><pre class="md">${esc(e.message)}</pre></div>`);
  }
})();
